### ISSUE TITLE
Cannot install from Flake due to cargo build fail

### ISSUE THREAD

**Issue by SimonBrandner** [rafaelrc7/wayland-pipewire-idle-inhibit#7]
I am trying to install from Flake as specified in https://github.com/rafaelrc7/wayland-pipewire-idle-inhibit#nix-flake-recommended, but it fails to build

```
❯ sudo nixos-rebuild --flake .#TuxedoInfinityBook14Gen8 switch --option eval-cache false
warning: Git tree '/home/simon/Data1/GIT/Other/dotfiles' is dirty
warning: updating lock file '/home/simon/Data1/GIT/Other/dotfiles/flake.lock':
• Updated input 'wayland-pipewire-idle-inhibit/nixpkgs':
    'github:nixos/nixpkgs/7bb2ccd8cdc44c91edba16c48d2c8f331fb3d856' (2024-04-25)
  → follows 'nixpkgs'
warning: Git tree '/home/simon/Data1/GIT/Other/dotfiles' is dirty
building the system configuration...
warning: Git tree '/home/simon/Data1/GIT/Other/dotfiles' is dirty
error: builder for '/nix/store/g4fk9in90ljmlli1glsm9lmicpsri95x-wayland-pipewire-idle-inhibit-0.5.1.drv' failed with exit code 101;
       last 10 log lines:
       > Finished cargoSetupPostPatchHook
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > Running phase: buildPhase
       > Executing cargoBuildHook
       > ++ env CC_X86_64_UNKNOWN_LINUX_GNU=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc CXX_X86_64_UNKNOWN_LINUX_GNU=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/c++ CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc CC_X86_64_UNKNOWN_LINUX_GNU=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc CXX_X86_64_UNKNOWN_LINUX_GNU=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/c++ CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc CARGO_BUILD_TARGET=x86_64-unknown-linux-gnu HOST_CC=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc HOST_CXX=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/c++ cargo build -j 8 --target x86_64-unknown-linux-gnu --frozen --profile release
       > error: package `zbus_macros v4.1.2` cannot be built because it requires rustc 1.75 or newer, while the currently active rustc version is 1.73.0
       > Either upgrade to rustc 1.75 or newer, or use
       > cargo update -p zbus_macros@4.1.2 --precise ver
       > where `ver` is the latest version of `zbus_macros` supporting rustc 1.73.0
       For full logs, run 'nix log /nix/store/g4fk9in90ljmlli1glsm9lmicpsri95x-wayland-pipewire-idle-inhibit-0.5.1.drv'.
error: 1 dependencies of derivation '/nix/store/5c5108pj9b96wbzvwjpsf1348pi0ab2z-wayland-pipewire-idle-inhibit.service.drv' failed to build
error: 1 dependencies of derivation '/nix/store/z6chvin57f4z2rvyyksy3wc24y3nsyy2-home-manager-files.drv' failed to build
error: 1 dependencies of derivation '/nix/store/dkbr2i67bdvqhxs7nv6zza8y4z3fikjn-home-manager-generation.drv' failed to build
error: 1 dependencies of derivation '/nix/store/6ls9s4d95j0gxmp71p3x5x3xpdl6cmvb-unit-home-manager-simon.service.drv' failed to build
error: 1 dependencies of derivation '/nix/store/jyk5xyrszg2348p8y1zrhwrx5y3kbf7r-system-units.drv' failed to build
error: 1 dependencies of derivation '/nix/store/c6j9kr3apq28dclpssp5l3lx9i0visdg-etc.drv' failed to build
error: 1 dependencies of derivation '/nix/store/a3pgzha5ap4qikk2pckw1afxny4pz2pi-nixos-system-Simon-s-Tuxedo-InfinityBook-14-Gen8-23.11.20240422.b500489.drv' failed to build
```

**Comment by rafaelrc7** [comment]
I believe the issue is that you are using nixos stable (23.11) and it seems that it has a rustc version that is too old (1.73.0) for the crate dependencies. You could try removing the 
```
inputs.nixpkgs.follows = "nixpkgs";
```
line from the `wayland-pipewire-idle-inhibit` flake input and then rebuilding. This will make the package to be built using the default nixos-unstable channel, and thus have a more up-to-date rustc version.

**Comment by SimonBrandner** [comment]
Hmm, I am instead getting this: 

```
❯ sudo nixos-rebuild --flake .#TuxedoInfinityBook14Gen8 switch --option eval-cache false
warning: Git tree '/home/simon/dotfiles' is dirty
warning: updating lock file '/home/simon/dotfiles/flake.lock':
• Updated input 'wayland-pipewire-idle-inhibit/nixpkgs':
    follows 'nixpkgs-unstable'
  → 'github:nixos/nixpkgs/58a1abdbae3217ca6b702f03d3b35125d88a2994' (2024-04-27)
warning: Git tree '/home/simon/dotfiles' is dirty
building the system configuration...
warning: Git tree '/home/simon/dotfiles' is dirty
error: builder for '/nix/store/g4fk9in90ljmlli1glsm9lmicpsri95x-wayland-pipewire-idle-inhibit-0.5.1.drv' failed with exit code 101;
       last 10 log lines:
       > Finished cargoSetupPostPatchHook
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > Running phase: buildPhase
       > Executing cargoBuildHook
       > ++ env CC_X86_64_UNKNOWN_LINUX_GNU=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc CXX_X86_64_UNKNOWN_LINUX_GNU=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/c++ CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc CC_X86_64_UNKNOWN_LINUX_GNU=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc CXX_X86_64_UNKNOWN_LINUX_GNU=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/c++ CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc CARGO_BUILD_TARGET=x86_64-unknown-linux-gnu HOST_CC=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc HOST_CXX=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/c++ cargo build -j 8 --target x86_64-unknown-linux-gnu --frozen --profile release
       > error: package `clap v4.5.4` cannot be built because it requires rustc 1.74 or newer, while the currently active rustc version is 1.73.0
       > Either upgrade to rustc 1.74 or newer, or use
       > cargo update -p clap@4.5.4 --precise ver
       > where `ver` is the latest version of `clap` supporting rustc 1.73.0
       For full logs, run 'nix log /nix/store/g4fk9in90ljmlli1glsm9lmicpsri95x-wayland-pipewire-idle-inhibit-0.5.1.drv'.
error: 1 dependencies of derivation '/nix/store/81j3s5rm2qwpipj6h0v28lqclab3plib-system-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/166fdxs7s1zzyr0y1mbkxbf1c39apjqi-nixos-system-Simon-s-Tuxedo-InfinityBook-14-Gen8-23.11.20240428.12430e4.drv' failed to build
```

**Comment by rafaelrc7** [comment]
This seems to be the same issue, it is trying to use rustc 1.73.0. Could you try adding nixos-unstable as an input, if you dont have it already, and set the wayland-pipewire-idle-inhibit flake to follow it?

**Comment by SimonBrandner** [comment]
That gives me the same thing as it did the first time (I think):

```
❯ sudo nixos-rebuild --flake .#TuxedoInfinityBook14Gen8 switch --option eval-cache false
warning: Git tree '/home/simon/dotfiles' is dirty
building the system configuration...
warning: Git tree '/home/simon/dotfiles' is dirty
error: builder for '/nix/store/g4fk9in90ljmlli1glsm9lmicpsri95x-wayland-pipewire-idle-inhibit-0.5.1.drv' failed with exit code 101;
       last 10 log lines:
       > Finished cargoSetupPostPatchHook
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > Running phase: buildPhase
       > Executing cargoBuildHook
       > ++ env CC_X86_64_UNKNOWN_LINUX_GNU=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc CXX_X86_64_UNKNOWN_LINUX_GNU=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/c++ CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc CC_X86_64_UNKNOWN_LINUX_GNU=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc CXX_X86_64_UNKNOWN_LINUX_GNU=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/c++ CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc CARGO_BUILD_TARGET=x86_64-unknown-linux-gnu HOST_CC=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/cc HOST_CXX=/nix/store/cb4mxc1mnl97590zbbhclgvqn4iqqap7-gcc-wrapper-12.3.0/bin/c++ cargo build -j 8 --target x86_64-unknown-linux-gnu --frozen --profile release
       > error: package `zvariant v4.0.2` cannot be built because it requires rustc 1.75 or newer, while the currently active rustc version is 1.73.0
       > Either upgrade to rustc 1.75 or newer, or use
       > cargo update -p zvariant@4.0.2 --precise ver
       > where `ver` is the latest version of `zvariant` supporting rustc 1.73.0
       For full logs, run 'nix log /nix/store/g4fk9in90ljmlli1glsm9lmicpsri95x-wayland-pipewire-idle-inhibit-0.5.1.drv'.
error: 1 dependencies of derivation '/nix/store/3mrvf192snk1m1p99d0zcwv20x9nzsv2-system-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/22i975nkkys3fhiby0wg050n8wkz9rhy-nixos-system-Simon-s-Tuxedo-InfinityBook-14-Gen8-23.11.20240428.12430e4.drv' failed to build
```

which I cannot for the life of me explain...

**Comment by rafaelrc7** [comment]
This is really strange... I'm not sure why it is not working. But I have another idea. The module has a option that lets you set the package, you could set it to use the stable package, it should only be a problem if you are trying to use the `dbus` feature, that is required for non-wayland or some DEs such as KDE.

You can do this:

```nix
  services.wayland-pipewire-idle-inhibit = {
    enable = true;
    ...
    package = pkgs.wayland-pipewire-idle-inhibit;
    ...
  };

```

**Comment by SimonBrandner** [comment]
The reason I want to use the flake is actually to have access to the features from unstable (dry-run mainly) :sweat_smile:

**Comment by rafaelrc7** [comment]
Ah, I see. In this case, instead of using `pkgs` from nixos-stable try using it from nixos-unstable. I believe there is some issues with the flake package dependency locking, but if you use the package from nixos-unstable, I believe it should work!

**Comment by rafaelrc7** [comment]
Yeah, removing the `inputs.nixpkgs.follows = "nixpkgs";` should have really worked... Can you copy and paste the snippets of the inputs that you tried and did not work?

**Comment by rafaelrc7** [comment]
Yeah, it really seems that it should work, I did the following test flake:

```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/23.11";
    flake-parts.url = "github:hercules-ci/flake-parts";

    wayland-pipewire-idle-inhibit = {
      url = "github:rafaelrc7/wayland-pipewire-idle-inhibit";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };

  outputs = inputs@{ wayland-pipewire-idle-inhibit, ... }:
    inputs.flake-parts.lib.mkFlake { inherit inputs; } {
      systems = [ "x86_64-linux" ];
      perSystem = { system, ... }: {
        packages.default = wayland-pipewire-idle-inhibit.packages.${system}.default;
      };
    };
}
```

Then, I tried running `nix build` and got the expected error stating that `error: package `clap_builder v4.5.2` cannot be built because it requires rustc 1.74 or newer, while the currently active rustc version is 1.73.0`.

I then removed the `inputs.nixpkgs.follows` from the `wayland-pipewire-idle-inhibit` input flake:

```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/23.11";
    flake-parts.url = "github:hercules-ci/flake-parts";

    wayland-pipewire-idle-inhibit.url = "github:rafaelrc7/wayland-pipewire-idle-inhibit";
  };

  outputs = inputs@{ wayland-pipewire-idle-inhibit, ... }:
    inputs.flake-parts.lib.mkFlake { inherit inputs; } {
      systems = [ "x86_64-linux" ];
      perSystem = { system, ... }: {
        packages.default = wayland-pipewire-idle-inhibit.packages.${system}.default;
      };
    };
}
```

And the, I ran `nix build` again and it worked!

```sh
$ nix build 
warning: updating lock file 'flake.lock':
• Updated input 'wayland-pipewire-idle-inhibit/nixpkgs':
    follows 'nixpkgs'
  → 'github:nixos/nixpkgs/58a1abdbae3217ca6b702f03d3b35125d88a2994?narHash=sha256-mdTQw2XlariysyScCv2tTE45QSU9v/ezLcHJ22f0Nxc%3D' (2024-04-27)
```

And the package was available in `result`.

**Comment by rafaelrc7** [comment]
Actually I think I found the reason? The last comment, that I hid and marked as outdated showed that when I removed the `inputs.nixpkgs.follows` the flake actually used my system's `nixpkgs` that is `nixos-unstable` and thats why it worked for me. When you removed it in yours it still followed your system's `nixpkgs` (23.11), so it makes no difference? I'm confused why that's the case, I believed that withouth the follows it would use it's own lock.

**Comment by rafaelrc7** [comment]
Ok, I believe I found an answer! It seems that if you remove a `input.nixpkgs.follows` you need to update the flake, in this case using:
```sh
nix flake update wayland-pipewire-idle-inhibit
```
Otherwise, it will default to your system nixpkgs lock (in my case unstable and yours 23.11, and thats why it was still using an old rustc version). I am not sure why you need this extra update step and seems to be a bug in the flake implementation. But still...

In summary, you can remove the `package = pkgs.wayland-pipewire-idle-inhibit`, you should remove the `input.nixpkgs.follows` from the `wayland-pipewire-idle-inhibit` flake input, run `$ nix flake update wayland-pipewire-idle-inhibit` and, finally, it should build.

**Comment by rafaelrc7** [comment]
@SimonBrandner Hello, did you manage to make it work with the previous suggestions?

**Comment by SimonBrandner** [comment]
Sorry, I managed to forget about my inbox tab for a while... I did try what you suggested but it did not work. I wanted to experiment a bit more but I ended up switching to unstable due to something else anyhow...

**Comment by rafaelrc7** [comment]
By this do you mean my last message about the `nix flake update wayland-pipewire-idle-inhibit`? Because this one worked when I tested it. However, you mentioned you switched to unstable, I believe that fixes the issues you were having, as the flake would point to unstable too, right?

**Comment by SimonBrandner** [comment]
Yep, your suggestion did not help sadly (can't see why...). Yes, switching to unstable fixed the issue

**Comment by SimonBrandner** [comment]
(not sure, if I should close)

**Comment by rafaelrc7** [comment]
Weird. yeah, as you are using unstable now, I believe this issue should be closed. If you run into this problem again, you can reopen it. I'll close it.
